### ISSUE TITLE
[RFC] Support Multiple Parameter Groups in FSDP

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#76382]
### Problem & Motivation
Currently, `FullyShardedDataParallel` (FSDP) flattens parameters into a single `FlatParameter`, losing the ability to manipulate at the granularity of the original unflattened parameters. This prevents users from defining multiple optimizer parameter groups, where unflattened parameters comprising the same `FlatParameter` span different parameter groups. This RFC addresses this problem of supporting multiple parameter groups.

### API
```
model: torch.nn.Module = ...
fsdp_model = FSDP(model, ...)
for ... in ...:                        # for each desired parameter group
    param_group: Dict[str, Any] = ...  # using original parameter variables from `model`
    fsdp_model.add_param_group(param_group)
param_groups: List[Dict[str, Any]] = fsdp_model.param_groups()
optim = torch.optim.Adam(param_groups, ...)
```

### `torch.optim.Optimizer`
We propose the API to closely follow that of `torch.optim.Optimizer` to minimize user dissonance.

To use multiple parameter groups `pg_1`, ..., `pg_k` with a `torch.optim.Optimizer`, there are two general workflows:
1. Initialize the torch.optim.Optimizer with `pg_1`, and iteratively add `pg_2`, ..., `pg_k` using `torch.optim.Optimizer.add_param_group()`.
2. Initialize the `torch.optim.Optimizer` with a `List[Dict[str, Any]]` consisting of `pg_1`, ..., `pg_k` directly.
3. Use an approach between 1 and 2 (i.e. initialize the `torch.optim.Optimizer` with `pg_1`, ..., `pg_m`, and iteratively add the remaining `k - m` using `torch.optim.Optimizer.add_param_group()`).
  
A few notable points:
- Each `add_param_group()` call performs a [disjointedness check](https://github.com/pytorch/pytorch/blob/master/torch/optim/optimizer.py#L286-L291) with each existing parameter group.
- It is the user's responsibility to account for all parameters to optimize.

### `FullyShardedDataParallel`

We introduce two APIs that work synergistically, following the iterative `add_param_group()` approach to multiple `torch.optim.Optimizer` parameter groups:
- `add_param_group(param_group: Dict[str, Any]) -> None`
- `param_groups() -> List[Dict[str, Any]]`

This avoids requiring changes to the FSDP constructor (e.g. adding a `param_groups` argument).

**`add_param_group()`**
-  _Arguments:_
`param_group_ (Dict[str, Any]):` The keys include `"params"` mapping to the original unflattened parameters and hyperparameter names mapping to their values. The method checks that the unflattened parameters in the group are not present in any existing parameter groups known to the FSDP instance.
- For generality, the **root** FSDP module will manage all parameters in its hierarchy (to follow the spirit of `root_module.parameters()`), meaning that `add_param_group()` should only be called on the root module and the argument can include original parameters corresponding to **any** FSDP module in the hierarchy.
- _Implementation:_
The FSDP constructor destructively modifies the user's original unflattened `torch.nn.Parameter`s `.data` variable to be a view into the corresponding `FlatParameter` and saves a reference to the parameter -- going forward, call these original but modified `torch.nn.Parameter`s that are still accessible to the user _flattened parameter views_.
Then, the FSDP module constructs a `dict` mapping from _flattened parameter view_ to a tuple (`FlatParameter`, offset), where the `FlatParameter` contains the parameter's data at the offset. When the root is set (in `_lazy_init()`), the root aggregates the `dict` from its children into its own `dict`.
The user constructs parameter groups using the _flattened parameter views_ and passes in the parameter group into `add_param_group()` (called on the root). The root may use its aggregated `dict` to bookkeep internally for the parameter group.
In `_finalize_params()` in `_wait_for_post_backward()` (only called on the root module), the root module uses its saved  _flattened parameter views_ to point each's `.grad` to the correct storage using the offset information in its saved `dict`. This ensures that the optimizer step uses the appropriate gradients.
_Addendum:_ We may also add a check to see if the `FlatParameter` storage changes (e.g. by the user) and if so, the root re-point s each _flattened parameter view_'s `.data` to the correct offset in the `FlatParameter`.
- Like for `torch.optim.Optimizer`, if the user does not include all original parameters in some `add_param_group()` call, then the accompanying `param_groups()` method will not include those parameters.

**`param_groups()`**
- _Returns:_
`param_groups (List[Dict[str, Any]]):` A `list` of the parameter groups, one for each `add_param_group()` call. Each parameter group's `"params"` key maps to the _flattened parameter view_ (which is the original `torch.nn.Parameter` variable) managed by the root FSDP module. The hyperparameter mappings are simply included unchanged.

There is no unflattening and re-flattening in this implementation. Everything is handled using views into the `FlatParameter`s.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by awgu** [comment]
Commenting to say that this RFC needs to be updated with new developments and discussion. I will update it some time after the branch cut (5/19). This RFC is also blocked by the other RFC https://github.com/pytorch/pytorch/issues/76501.

**Comment by awgu** [comment]
Closing due to more internal discussion.
